### PR TITLE
feat: image plugin take arbitrary url

### PR DESCRIPTION
**Pull request by ivan-ngchakming**
### Description

Added modal for image plugin to allow option to add custom URI and files. 

~~File button is disabled, to be implemented later (or should I do it in this PR too?).~~
Upload file feature to be implemented in another PR.

part of #1799

### Screenshots
image

image

image

### Issues

Encountered a few issues, not sure if should be addressed in this PR.

1. Image cannot be added if current selection is null.
2. Random image added using https://source.unsplash.com/random are cached by `useSuspenseImage`.

### CODE REVIEW

File: packages/lexical-playground/src/plugins/ToolbarPlugin.jsx [hunk 1]
```diff
@@ -273,6 +275,53 @@ function FloatingLinkEditor({editor}: {editor: LexicalEditor}): React$Node {
...
+    onClose();
+  };
+
+  const isDisabled = text === '';
+
+  return (
+    <>
+      {!mode && (
+        <div className="ToolbarPlugin__dialogImageActions">
+          <Button onClick={onClick()}>Sample</Button>
```
Comment (zurfyx) [inline packages/lexical-playground/src/plugins/ToolbarPlugin.jsx:299]:
```suggestion
          <Button onClick={() => onClick({})}>Sample</Button>
```

File: packages/lexical-playground/src/plugins/ToolbarPlugin.jsx [hunk 2]
```diff
@@ -273,6 +275,53 @@ function FloatingLinkEditor({editor}: {editor: LexicalEditor}): React$Node {
...
 
+function InsertImageDialog({
+  activeEditor,
+  onClose,
+}: {
+  activeEditor: LexicalEditor,
+  onClose: () => void,
+}): React$Node {
+  const [text, setText] = useState('');
+  const [mode, setMode] = useState<'url' | 'file' | ''>('');
```
Comment (zurfyx) [inline packages/lexical-playground/src/plugins/ToolbarPlugin.jsx:286]:
```suggestion
  const [mode, setMode] = useState<null | 'url' | 'file'>(null);
```

File: packages/lexical-playground/src/plugins/ToolbarPlugin.jsx [hunk 3]
```diff
@@ -273,6 +275,53 @@ function FloatingLinkEditor({editor}: {editor: LexicalEditor}): React$Node {
...
+
+  const onClick = (payload) => () => {
+    activeEditor.dispatchCommand(INSERT_IMAGE_COMMAND, payload);
```
Comment (zurfyx) [inline packages/lexical-playground/src/plugins/ToolbarPlugin.jsx:289]: Let's type payload properly in for INSERT_IMAGE_COMMAND (check the generic type of the command at the top of the ImagesPlugin file)

File: packages/lexical-playground/src/plugins/ToolbarPlugin.jsx [hunk 4]
```diff
@@ -273,6 +275,53 @@ function FloatingLinkEditor({editor}: {editor: LexicalEditor}): React$Node {
...
+          <div className="ToolbarPlugin__dialogActions">
+            <Button disabled={isDisabled} onClick={onClick({src: text})}>
+              Confirm
+            </Button>
+          </div>
+        </>
+      )}
+    </>
+  );
+}
```
Comment (zurfyx) [inline packages/lexical-playground/src/plugins/ToolbarPlugin.jsx:323]: Can we move this to a separate component even if it's within the same file? Having a `isDisabled` at the top of the file is a bit confusing as you don't whether this is referring to the first or second part of the modal

File: packages/lexical-playground/src/plugins/ToolbarPlugin.jsx [hunk 5]
```diff
@@ -916,7 +965,13 @@ export default function ToolbarPlugin(): React$Node {
             </button>
             <button
               onClick={() => {
-                activeEditor.dispatchCommand(INSERT_IMAGE_COMMAND);
+                // activeEditor.dispatchCommand(INSERT_IMAGE_COMMAND);
```
Comment (zurfyx) [inline packages/lexical-playground/src/plugins/ToolbarPlugin.jsx:968]:
```suggestion
```

**Review by zurfyx — COMMENTED** [review]
LGTM, thanks for doing this! Let's handle File separately as you suggested. Added a few nits but they're just nits

On a separate note, there's a few E2E tests that depend on the current UI flow for images. You will have to update them to match the new flow. I would suggest you create a helper function for the Sample image case specifically since it's a 3 step process and can likely save us from some good repetition

File: packages/lexical-playground/src/plugins/ToolbarPlugin.css [hunk 6]
```diff
@@ -12,3 +12,14 @@
   justify-content: right;
   margin-top: 20px;
 }
+
+.ToolbarPlugin__dialogImageActions {
```
Comment (thegreatercurve) [inline packages/lexical-playground/src/plugins/ToolbarPlugin.css:16]: Just making it a little more general as I think we'll use this styling again:

```suggestion
.ToolbarPlugin__dialogButtonsList {
```

File: packages/lexical-playground/src/plugins/ImagesPlugin.js [hunk 7]
```diff
@@ -34,15 +37,15 @@ export default function ImagesPlugin(): React$Node {
...
         const selection = $getSelection();
         if ($isRangeSelection(selection)) {
           if ($isRootNode(selection.anchor.getNode())) {
             selection.insertParagraph();
           }
           const imageNode = $createImageNode(
-            yellowFlowerImage,
-            'Yellow flower in tilt shift lens',
+            payload?.src || yellowFlowerImage,
+            payload?.altText || 'Yellow flower in tilt shift lens',
```
Comment (thegreatercurve) [inline packages/lexical-playground/src/plugins/ImagesPlugin.js:48]: For accessibility reasons, I'd make both of these required type arguments, so we don't potentially add the wrong above alt text to any image URL from Unsplash, if only the `src` is supplied.
```suggestion
            payload.src ,
            payload.altText,
```

File: packages/lexical-playground/src/plugins/ToolbarPlugin.jsx [hunk 8]
```diff
@@ -273,6 +275,53 @@ function FloatingLinkEditor({editor}: {editor: LexicalEditor}): React$Node {
...
+      )}
+      {mode === 'url' && (
+        <>
+          <Input
+            label="Image URL"
+            placeholder="i.e. https://source.unsplash.com/random"
+            onChange={setText}
+            value={text}
+          />
+          <div className="ToolbarPlugin__dialogActions">
```
Comment (thegreatercurve) [inline packages/lexical-playground/src/plugins/ToolbarPlugin.jsx:314]: Could we maybe add an Input for alt text as well?

**Review by thegreatercurve — COMMENTED** [review]
Just a couple of things, but looks good to me! Thanks for contributing!

File: packages/lexical-playground/__tests__/utils/index.mjs [hunk 9]
```diff
@@ -525,6 +525,14 @@ export async function insertTable(page) {
   );
 }
 
+export async function insertSampleImage(page) {
+  await selectFromInsertDropdown(page, '.image');
+  await click(
+    page,
+    '.ToolbarPlugin__dialogButtonsList > .Button__root >> nth=0',
```
Comment (zurfyx) [inline packages/lexical-playground/__tests__/utils/index.mjs:532]: I like the fact you introduced this utility function! Can we add data-test-id to the button? Looks somewhat fragile for a utils function and would also simplify this

**Review by zurfyx — APPROVED** [review]
Thank you, just one last nit if you don't mind

**Review by zurfyx — APPROVED** [review]

**Pull request merged**